library/core/src/macros/mod.rs

```diff
@@ -137,7 +137,7 @@ macro_rules! assert_ne {
 /// assert_matches!(c, Ok(x) | Err(x) if x.len() < 100);
 /// ```
 #[unstable(feature = "assert_matches", issue = "82775")]
-#[allow_internal_unstable(core_panic)]
+#[allow_internal_unstable(core_panic, const_format_args)]
 #[rustc_macro_transparency = "semitransparent"]
 pub macro assert_matches {
```

Comment:
So... my worry about forward compatibility is that I believe one could now call `assert_matches` on a user defined type that does not implement `const Display`. If the match succeeds, there will be no error, even if we may want to error in the future because there is no `const Display` impl.

For the assert macros that's not so much a problem, as they already require `const PartialEq` essentially.

I'm also not sure how this affects the `panic!` macro. Can we now do `if cond { panic!("{}", some_user_defined_type) }`? If so, I think that will bite us in the future.

Reply:
> I'm also not sure how this affects the `panic!` macro. Can we now do `if cond { panic!("{}", some_user_defined_type) }`? If so, I think that will bite us in the future.

Yes. But I don't think this will bite us in the future. We can just display "<cannot be formatted>" for types that didn't have `const Display` implementation (could be achieved either by specialization on whether a const Display impl exists, or an intrinsic/lang_item).

IMO panicking with some info is better than not allowing the user to panic; user would just remove information from panic messages to those accepted by const eval, and then the info that could have been displayed in runtime would be lost.

Reply:
Well... the same argument goes the other way. Users may want to know when we are ignoring the code they wrote because they missed a `const` somewhere.

Like... it feels to me like allowing `panic!("{}", non_display_type)` at runtime and just printing "sorry can't do" when panicking.

It would also mean we would have to keep the fallback logic forever, which seems unfortunate when we could just force users to use const Display impls at some point.

Reply:
> Well... the same argument goes the other way. Users may want to know when we are ignoring the code they wrote because they missed a `const` somewhere.
> 
> Like... it feels to me like allowing `panic!("{}", non_display_type)` at runtime and just printing "sorry can't do" when panicking.

Well, this all happens at compile time, so (1) it's easy to reproduce for the user (no need to run the program and wait for the panic to hit), so less info is acceptable than if it happens in the runtime and (2) the compiler has a lot of info to produce useful diagnostics. The compiler could produce additional diagnostics saying `<ty> cannot be formatted because its Display impl is not const`.

> 
> It would also mean we would have to keep the fallback logic forever, which seems unfortunate when we could just force users to use const Display impls at some point.

It wouldn't be a lot or a complex logic though. And honestly I don't think we are anywhere near of getting to a point that we can have a working (and ergnomical)  `const Display` any time soon. Like whether `#[derive(Debug)]` would implement `const Debug` or just `Debug`? With the current approach I can just eval all derived Debug, and if any field isn't const formattable it would only need a diagnostics.

And then there is an issue that people want more powerful panic right now, e.g. with `Result::unwrap`. If we want to reserve the possibily of removing the fallback logic, what would the bound look like? Solution would most likely involve an unstable `ConstDebug` bound and then it could probably never be stablised.

Reply:
> Well, this all happens at compile time, so (1) it's easy to reproduce for the user (no need to run the program and wait for the panic to hit), so less info is acceptable than if it happens in the runtime 

Not really: the panic might be in a `const fn` in a library, so while it runs at compile time it does not necessarily run at compile time *of the current crate*.

Reply:
Opened a zulip thread in https://rust-lang.zulipchat.com/#narrow/stream/219381-t-libs/topic/const.20panic!.20for.20all.20types.20.2390488 so we can discuss this with the libs API team, as I am unsure of how to proceed here.